README.md
Chesstron
===================


Chess player AI written in JavaScript

Sept 24, 2014 - Update: Added JavaScript Workers!

Try the demo at:
http://jacklehamster.github.com/chesstron/


About the project:
------------------

The idea is to develop a JavaScript based AI that will run in a sandboxed environment to decide the best move for a chess game.
Future plan is to have AI play vs each other ;-P

Setup your AI:
- Each AI unit is derived from a class Player (Look at ThinkingPlayer or RandomPlayer inside chessplayer.as for model)
- The Player has a member function "decide", which will be called every time a player needs to decide a move. The function takes a model and a callback.
- The model object can "make a move" and duplicate itself. It also has information about the pieces on the board.
A classic way to program AI is to duplicate the model, make it play various move, and determine based on the situation if
a move is realistic or not (for yourself, your move should end up in a position of winning. For the opponent, the move should
make you lose). To evaluate a board, the ThinkingPlayer algorithm simply attributes some values to pieces, and counts how
many of them are left (Queens are worth much more than other pieces).
- The Player is expected to take a long time to decide. To avoid javascript to get stuck, set the useWorker member variable
to the name of your class. Also make sure your class is in a file imported by chessworker.js.
- AI should not use Timer functions or native Random functions. There is a seeded Random function available. The reason why
we prefer a seeded Random function is because the plan is to have matches between AI. Each match will have a different seed,
but each match with a particular ssed should always end with the same result between two AI, no matter what.

This project is in progress. Right now there are two AI player. One that plays randomly (does not use Worker). Another one,
called ThinkingPlayer, that thinks using the classical AI approach described above. 
By default, the ThinkingPlayer is available. 

Feel free to give me comments.


History.txt
v1.0.0
This is the first history file, so it starts with everything.
- Chess Player with AI in JavaScript
- JavaScript workers
- Multiplayer ability

chessutil.js
/*
 
 Copyright (C) 2013 Vincent Le Quang
 
 This program is free software; you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation;
 either version 2 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along with this program;
 if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 
 Contact : vincentlequang@gmail.com
 
 */


function random() {
    return random.x = parseFloat('0.'+Math.sin(random.x).toString().substr(6));
}
random.x = 1000;

function getSearchParameters() {
    var query = document.location.search.substr(1);
    var querySplit = query.split("&");
    var obj = {};
    for(var i=0;i<querySplit.length;i++) {
        var q = querySplit[i].split("=");
        if(q[0] && q[1]) {
            obj[q[0]] = q[1];
        }
    }
    return obj;    
}

chessview.js
var emptyDataURI = "data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==";

var view = {
    draggedImage:null
};

var playerID = null;
var hoveredCell = null;
var draggedImage = null;
var selected = null;

function initView() {
    var chessboard = document.getElementById('chessboard');
    chessboard.id = "chessboard";
    
    var computerPlayDiv = chessboard.appendChild(document.createElement("div"));
    computerPlayDiv.id = "computerPlayDiv";
    var checkbox = computerPlayDiv.appendChild(document.createElement("input"));
    checkbox.id = "computerplayercheck";
    checkbox.type ="checkbox";
    checkbox.onclick = checkboxClick;
    computerPlayDiv.appendChild(document.createTextNode(" black played by computer"));
    
    var peerPlayDiv = chessboard.appendChild(document.createElement("div"));
    peerPlayDiv.id = "peerPlayDiv";
    peerPlayDiv.style.display = "none";
    
    chessboard.appendChild(document.createElement("br"));
    
    var table = chessboard.appendChild(document.createElement('table'));
    table.id = "chessboard-table";
    table.style.backgroundColor = "#222222";
    table.style.border = "1px solid black";
    var tbody = table.appendChild(document.createElement('tbody'));
    
    //  header
    var letters = "ABCDEFGH";
    var tr = tbody.appendChild(document.createElement('tr'));
    var td = tr.appendChild(document.createElement('th'));
    for(var col=0;col<8;col++) {
        var th = tr.appendChild(document.createElement('th'));
        th.height = 20;
        th.style.textAlign = "center";
        th.innerHTML = letters.charAt(col).fontcolor("silver");
    }
    var th = tr.appendChild(document.createElement('th'));
    th.width = 20;
    th.height = 20;
    
    //  history blocks
    var history = tr.appendChild(document.createElement('td'));
    history.id = "history";
    history.width = "100px";
    history.style.padding = "20px";
    history.style.backgroundColor = "#EEEEEE";
    history.style.textAlign = "center";
    history.style.verticalAlign = "top";
    history.setAttribute("rowspan",10);
    
    for(var row=0;row<8;row++) {
        var tr = tbody.appendChild(document.createElement('tr'));
        var th = tr.appendChild(document.createElement('th'));
        th.style.textAlign = "center";
        th.style.width = 20;
        th.innerHTML = (row+1).toString().fontcolor("silver");
        for(var col=0;col<8;col++) {
            var td = tr.appendChild(document.createElement('td'));
            td.style.width = "50px";
            td.style.height = "50px";
            td.align = "center";
            td.setAttribute("background",(row+col)%2==0?"img/darkcell.png":"img/lightcell.png");
            var div = td.appendChild(document.createElement("div"));
            div.id = "div"+col+"_"+row;
            var img = div.appendChild(document.createElement("img"));
            img.id = "cell"+col+"_"+row;
            img.src = emptyDataURI;
            img.style.width = "44";
            img.style.height = "44";
            img.position = {x:col,y:row};
            img.onmouseover = cellHover;
            img.onmouseout = cellOut;
            img.onmousedown = cellDown;
            img.draggable = false;
            img.style.border = "3px solid transparent";
        }
        var th = tr.appendChild(document.createElement('th'));
        th.style.width = 20;
        th.style.height = 20;
    }
    
    
    //  footer
    var tr = tbody.appendChild(document.createElement('tr'));
    var th = tr.appendChild(document.createElement('th'));
    for(var col=0;col<8;col++) {
        var th = tr.appendChild(document.createElement('th'));
        th.width = 20;
        th.height = 20;
    }
    var th = tr.appendChild(document.createElement('th'));
    th.width = 20;
    th.height = 20;
    
    chessboard.appendChild(document.createElement("br"));
    
    
    var div = chessboard.appendChild(document.createElement("div"));
    div.id = "divturn";
    
    var button = chessboard.appendChild(document.createElement("button"));
    button.id = "startButton";
    button.innerHTML = "Start game";
    button.style.width = 100;
    button.style.height = 50;
    button.onclick = startGameClick;
}

function setDraggedImageURL(url,x,y) {
    if(url) {
        if(!view.draggedImage) {
            view.draggedImage = document.body.appendChild(document.createElement("img"));
            view.draggedImage.style.position = "absolute";
            view.draggedImage.style.pointerEvents = "none";
            view.draggedImage.draggable = false;
        }
        view.draggedImage.src = url;
        view.draggedImage.style.left = view.draggedImage.style.posLeft = x-25;
        view.draggedImage.style.top = view.draggedImage.style.posTop = y-25;
    }
    else if(view.draggedImage) {
        document.body.removeChild(view.draggedImage);
        view.draggedImage = null;
    }
}

function hoverCell(x,y,hilight) {
    hoveredCell = getViewCell(x,y);
    for(var row=0;row<8;row++) {
        for(var col=0;col<8;col++) {
            var viewCell = getViewCell(col,row);
            viewCell.style.borderColor = hilight && viewCell==hoveredCell ? "#00FF00" : "transparent";
        }
    }
}

function selectCell(x,y) {
    selected = x>=0 && y>=0 ? {x:x,y:y} : null;
    for(var row=0;row<8;row++) {
        for(var col=0;col<8;col++) {
            var div = getDiv(col,row);
            div.css({opacity:x==col&&y==row?.2:1});
        }
    }
}

function startGameClick(e) {
    if(onStartGame) {
        onStartGame();
    }
}

function cellOut() {
    if(onCellOut) {
        onCellOut(this.position.x,this.position.y);
    }
}

function cellHover() {
    if(onCellHover) {
        onCellHover(this.position.x,this.position.y);
    }
}

function cellDown() {
    if(onCellDown) {
        onCellDown(this.position.x,this.position.y);
    }
}

function checkboxClick() {
    if(onCheckComputerPlayer) {
        onCheckComputerPlayer(isComputerPlaying());
    }
}

function getDiv(col,row) {
    return $("#div"+col+"_"+row);
}

function getViewCell(col,row) {
    return $("#cell"+col+"_"+row)[0];
}

function updateView(model) {
    for(var row=0;row<8;row++) {
        for(var col=0;col<8;col++) {
            var piece = model.getCell(col,row);
            var pieceURL = getCachedImageURL(piece);
            var viewCell = getViewCell(col,row);
            viewCell.src = pieceURL;
        }
    }
    document.getElementById("divturn").innerHTML =
        model.checkMate() ? (model.turn=="w" ? "check mate. white lost" : "check mate. black lost") :
        model.inCheck() ? (model.turn=="w" ? "check! white's turn" : "check! black's turn") :
        model.turn=="w" ? "white's turn" :
        model.turn=="b" ? "black's turn" : "";
    document.getElementById("startButton").style.display = model.turn ? "none" : "";
    var history = document.getElementById("history");
    history.innerHTML = "";
    if(model.history.length) {
        var undoButton = (history.appendChild(document.createElement("div"))).appendChild(document.createElement('button'));
        undoButton.appendChild(document.createTextNode("undo"));
        undoButton.style.display = isPeerPlaying() ? "none":"";
        undoButton.onclick = function() {
            if(onUndo) {
                onUndo();
            }
        };
    }
    for(var i=0;i<model.history.length;i++) {
        var move = model.history[i];
        var div = history.appendChild(document.createElement("div"));
        div.innerHTML = displayMove(move,i%2==0?"#999999":"black");
    }
    if(model.turn && !playerID) {
        playerID = model.players[model.turn];
    }
}

function isComputerPlaying() {
    return document.getElementById("computerplayercheck").checked;
}

function setPeerPlaying(server,room) {
    computerPlayDiv.style.display = "none";
    var js = document.createElement("script");
    js.type = "text/javascript";
    js.src = server + "?action=enter&javascript=1&room=" + room;
    document.head.appendChild(js);
    
    var peerPlayDiv = document.getElementById("peerPlayDiv");
    peerPlayDiv.style.display = "";
    peerPlayDiv.innerHTML = "Joining room <b>"+room+"</b> on "+server;
}

function isPeerPlaying() {
    var peerPlayDiv = document.getElementById("peerPlayDiv");
    return peerPlayDiv.style.display=="";
}

function updateServerMessage(message) {
    var peerPlayDiv = document.getElementById("peerPlayDiv");
    peerPlayDiv.innerHTML = message;
}

function setComputerPlaying(value) {
    document.getElementById("computerplayercheck").checked = value;
    checkboxClick();
}

function displayMove(move,color) {
    var letters = "ABCDEFGH";
    var moveStr = (letters.charAt(move[0].x)+(move[0].y+1)) + "-" +
        (letters.charAt(move[1].x)+(move[1].y+1));
    
    return moveStr.fontcolor(color);
}

function hilight(moves,target,red) {
    var moveHash = {};
    if(moves) {
        for(var i=0;i<moves.length;i++) {
            moveHash[moves[i].x+"_"+moves[i].y] = true;
        }
    }
    var targetCell = target ? getViewCell(target.x,target.y) : null;
    for(var row=0;row<8;row++) {
        for(var col=0;col<8;col++) {
            var viewCell = getViewCell(col,row);
            var hilightCell = moveHash[col+"_"+row];
            viewCell.style.borderColor = 
                hilightCell && targetCell==viewCell ? (red ? "#FF0000":"#FFFF00") :
                hilightCell ? "#FFBB00" : "transparent";
        }
    }
}

function onCosmo(update) {
    console.log(">>",update);
}

var cachedImages = {};

function getCachedImageURL(piece) {
    if(!cachedImages[piece]) {
        cachedImages[piece] = new Image();
        cachedImages[piece].src = getImageURL(piece);
    }
    return cachedImages[piece].src;
}

function getImageURL(piece) {
    switch(piece) {
        case "br": return "img/black-rook.svg"; break;
        case "bh": return "img/black-knight.svg"; break;
        case "bb": return "img/black-bishop.svg"; break;
        case "bq": return "img/black-queen.svg"; break;
        case "bk": return "img/black-king.svg"; break;
        case "bp": return "img/black-pawn.svg"; break;
        case "wr": return "img/white-rook.svg"; break;
        case "wh": return "img/white-knight.svg"; break;
        case "wb": return "img/white-bishop.svg"; break;
        case "wq": return "img/white-queen.svg"; break;
        case "wk": return "img/white-king.svg"; break;
        case "wp": return "img/white-pawn.svg"; break;
        default: return emptyDataURI;
    }
}


chessdebug.js
/*
 
 Copyright (C) 2013 Vincent Le Quang
 
 This program is free software; you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation;
 either version 2 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along with this program;
 if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 
 Contact : vincentlequang@gmail.com
 
 */

function ddd(n) {
    return !n || n<=0 ? [players.b.getCurrentChoice(this.model)] : players.b.spread(ddd(n-1));
}

function d(n) {
    setComputerPlaying(true);
    var choices = ddd(n);
    return choices;
}

chessplayer.js
/*
 
 Copyright (C) 2013 Vincent Le Quang
 
 This program is free software; you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation;
 either version 2 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along with this program;
 if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 
 Contact : vincentlequang@gmail.com
 
 */


var choiceLimit = 1000;
var pieceValues = {
    k:100000,
    q:10,
    r:5,
    b:3,
    h:3,
    p:1
};

if(typeof(console)!='undefined') {
    console.log("choiceLimit:",choiceLimit);
    console.log("pieceValues:",JSON.stringify(pieceValues,null,'\t'));
}

function Player() {
}

Player.prototype = {
    callback : null,
    worker: null,
    localHuman : false,
    useWorker: false,
    decide : function(model,callback) {
    },
    workerDecide : function(model,callback) {
        this.callback = callback;
        if(!this.worker) {
            this.worker = new Worker("chessworker.js");
        }
        this.worker.onmessage = function(oEvent) {
            callback(oEvent.data);
        };
        this.worker.postMessage(
            {
                worker : this.useWorker,
                model: {
                    wCastlingRight : model.wCastlingRight,
                    wCastlingLeft : model.wCastlingLeft,
                    bCastlingRight : model.bCastlingRight,
                    bCastlingLeft : model.bCastlingLeft,
                    history : [],
                    turn : model.turn,
                    board : model.board,
                    players : model.players
                },
                decide : this.decide.toString()
            }
        );
    }
};


//  RandomPlayer : Player

function RandomPlayer() {
}
RandomPlayer.prototype = new Player();
RandomPlayer.prototype.decide = function(model,callback) {
    var moves = model.getPossibleMoves();
    if(moves.length) {
        var moveIndex = Math.floor(random()*moves.length);
        callback(moves[moveIndex]);
    }
    else {
        callback(null);
    }
};

//  ThinkingPlayer : Player

function Choice(parent,player,model,move) {
    this.parent = parent;
    this.player = player;
    this.move = move;
    this.model = model;
    this.children = null;
    this.level = parent ? parent.level+1 : 0;
    this.value = player.evaluate(this.model);
    this.history = (this.parent ? this.parent.history: []).concat(this.move ?[getMoveId(this.move)]:[]);
}

Choice.prototype = {
    getTurn : function() {
        return this.model.turn;
    },
    spread : function() {
        var choices = [];
        if(this.model.hasKing()) {
            var moves = this.model.getPossibleMoves();
            for(var i=0;i<moves.length;i++) {
                var model = new ChessBoard(this.model);
                model.move(moves[i][0],moves[i][1],true,false);
                var choice = new Choice(this,this.player,model,moves[i]);
                choices.push(choice);
            }
            this.children = choices;
        }
        return choices;
    },
    clearEvaluation : function() {
        delete this.evaluation;
        if(this.children) {
            for(var i=0;i<this.children.length;i++) {
                this.children[i].clearEvaluation();
            }
        }
    },
    getRoot : function() {
        return this.isRoot() ? this : this.parent.getRoot();
    },
    isRoot : function() {
        return !this.parent;
    },
    evaluate : function() {
        if(this.isRoot()) {
            this.clearEvaluation();
        }
        var evaluateHelper = function(choice) {
            var children = choice.children;
            if(children) {
                for(var i=0;i<children.length;i++) {
                    if(children[i].evaluation===undefined) {
                        evaluateHelper(children[i]);
                    }
                }
                
                children.sort(compareChoices);
                if(choice.model.turn=='w')
                    children.reverse();
            }
            choice.evaluation = !children ? choice.value :
                children.length ? children[0].evaluation :
                !choice.model.inCheck()?0:
                choice.model.turn=='w'?-100000:
                choice.model.turn=='b'?+100000:
                0;
        };
        evaluateHelper(this);
    },
    siblings : function() {
        return this.parent ? this.parent.children : null;
    },
    deleteBranch : function() {
        if(this.parent && this.parent.children) {
            var index = this.parent.children.indexOf(this);
            if(index>=0) {
                this.parent.children.splice(index,1);
            }
            this.parent = null;
        }
        this.markBranch("deleted",true);
    },
    markBranch : function(property,value) {
        this[property] = value;
        if(this.children) {
            for(var i=0;i<this.children.length;i++) {
                this.children[i].markBranch(property,value);
            }
        }
    }
}

function compareChoices(choice1,choice2) {
    return choice1.evaluation-choice2.evaluation;
}

var timeout = null;

function ThinkingPlayer() {
    this.imagining = false;
    this.filtering = false;
    this.depth = 0;
    this.choices = null;
    this.useWorker = "ThinkingPlayer";
}
ThinkingPlayer.prototype = new Player();
ThinkingPlayer.prototype.decide = function(model,callback) {
    var moves = model.getPossibleMoves();
    if(moves.length) {
        this.think(model,callback);
    }
    else {
        callback(null);
    }
};
ThinkingPlayer.prototype.spread = function(choices) {
    var newChoices = [], hash = {};
    for(var i=0;i<choices.length;i++) {
        newChoices = newChoices.concat(choices[i].spread());
    }
    for(var i=newChoices.length-1;i>=0;i--) {
        if(!hash[newChoices[i].model.board]) {
            hash[newChoices[i].model.board] = true;
        }
        else {
            newChoices[i] = newChoices[newChoices.length-1];
            newChoices.pop();
        }
    }
    return newChoices;
};

ThinkingPlayer.prototype.getCurrentChoice = function(model) {
    return new Choice(null,this,model,null);
}

ThinkingPlayer.prototype.think = function(model,callback) {
    
    var root = this.getCurrentChoice(model);
    var choices = [root];
    var depth = 0;  //  how deep is the imagination

    do {
        while(choices.length<=choiceLimit) {
            choices = this.imagine(choices,1);
            depth ++;
        }
        
        while(choices.length>choiceLimit && root.children.length>1) {
            choices = this.filter(choices);
            root = choices[0].getRoot();
        }
    } while(root.children.length>1);

    callback(root.children[0].move);
}

ThinkingPlayer.prototype.imagine = function(choices,depth) {
    return depth<=0 ? choices : this.imagine(this.spread(choices),depth-1);
}

ThinkingPlayer.prototype.filter = function(choices) {
    var root = choices[0].getRoot();
    root.evaluate();
    if(root.children.length>1) {
        root.children[root.children.length-1].deleteBranch();
        for(var i=choices.length-1;i>=0;i--) {
            if(choices[i].deleted) {
                choices[i] = choices[choices.length-1];
                choices.pop();
            }
        }
    }
    return choices;
}

ThinkingPlayer.prototype.evaluate = function(model) {
    var value = 0;
    for(var row=0;row<8;row++) {
        for(var col=0;col<8;col++) {
            var cell = model.getCell(col,row);
            var dir = 0;
            if(cell.charAt(0)=='w') {
                dir = 1;
            }
            else if(cell.charAt(0)=='b'){
                dir = -1;
            }
            if(dir) {
                value += pieceValues[cell.charAt(1)] * dir;
            }
        }
    }
    return value;
}

//  HumanPlayer : Player

function HumanPlayer() {
}
HumanPlayer.prototype = new Player();
HumanPlayer.prototype.decide = function(model,callback) {
    this.callback = callback;
}
HumanPlayer.prototype.localHuman = true;

//  WorkerPlayer : Player

function WorkerPlayer() {
}
WorkerPlayer.prototype = new Player();
WorkerPlayer.prototype.decide = function(model,callback) {
    
}



chessworker.js
/*
 
 Copyright (C) 2013 Vincent Le Quang
 
 This program is free software; you can redistribute it and/or modify it under the terms of the
 GNU General Public License as published by the Free Software Foundation;
 either version 2 of the License, or (at your option) any later version.
 
 This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 See the GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along with this program;
 if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 
 Contact : vincentlequang@gmail.com
 
 */

importScripts("chessmodel.js","chessplayer.js","chessutil.js");

onmessage = function (oEvent) {
    var model = new ChessBoard(oEvent.data.model);
    var player = new (eval(oEvent.data.worker))();
    player.decide = eval("("+oEvent.data.decide+")");
    postMessage("TET123");
    player.decide(model,postMessage);
};

cosmo.php
<?php

    /*
     
     Copyright (C) 2013 Vincent Le Quang
     
     This program is free software; you can redistribute it and/or modify it under the terms of the
     GNU General Public License as published by the Free Software Foundation;
     either version 2 of the License, or (at your option) any later version.
     
     This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
     without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     See the GNU General Public License for more details.
     
     You should have received a copy of the GNU General Public License along with this program;
     if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
     
     Contact : vincentlequang@gmail.com
     
     */
    
    
   define("MEMDIR", "memoku");
   $action = $_REQUEST['action'];

   switch($action) {
      case "destroy":
          $room = $_REQUEST['room'];
          if($room) {
             $channel = md5($room);
             $channeldir = MEMDIR."/$channel";
             destroy($channeldir,true);
          }
          break;
      case "enter":
         $room = $_REQUEST['room'];
         $channel = md5($room);
         $memdir = MEMDIR;
         $channeldir = "$memdir/$channel";
         $javascript = $_REQUEST['javascript'];         

         if(!file_exists($channeldir) || !file_exists("$channeldir/data.json")) {
                if(!file_exists(MEMDIR)) {
                    mkdir(MEMDIR);
                }
                if(!file_exists($channeldir)) {
                    mkdir($channeldir);
                }
                if(!file_exists("$channeldir/1.json"))
                    file_put_contents("$channeldir/1.json","");
                if(!file_exists("$channeldir/data.json"))  {
                    $data = "1,{\"room\":\"$room\"}";
                    file_put_contents("$channeldir/data.json",$data);
                }
         }
         else {
             $data = file_get_contents("$channeldir/data.json");
         }
         $server = $_SERVER["HTTP_HOST"];
         $urlaccess = "http://$server/cosmo.php";
         $info = "{\"server\":\"$server\",\"room\":\"$room\",\"channel\":\"$channel\",\"memdir\":\"$memdir\",\"data\":[$data],\"url\":\"$urlaccess\"}";
         if(!$javascript) {
              header("Content-Type:text/plain");
              echo $info;
         }
         else { 
              header("Content-type: text/javascript");
              echo "var info=$info;";
?>

if(!window.cosmo) {

   function Cosmo(url) {
       this.spots = {};
       this.url = url;
   }
   Cosmo.prototype = {
       getSpot : function(roomname) {
           return this.spots[roomname] ? this.spots[roomname] : this.spots[roomname] = new Spot(this,roomname);
       }
   };

   function Spot(cosmo,roomname) {
      this.cosmo = cosmo;
      this.roomname = roomname;
   }

   Spot.prototype = {
      setInfo : function(channel,data,count) {
          this.channel = channel;
          this.data = data;
          this.count = count;
      },
      getNewScript : function() {
         var js = document.getElementById("script_"+this.roomname);
         if(js)
            document.head.removeChild(js);
         js = document.createElement("script");
         js.id = "script_"+this.roomname;
         js.type = "text/javascript";
         document.head.appendChild(js);
         return js;
      },
      loadScript : function() {
         var js = this.getNewScript();
         js.src = this.cosmo.url + "?" +
	      "action=get" +
              "&callback=cosmo.getSpot('" + this.roomname + "').update" +
	      "&channel=" + this.channel +
	      "&count=" + this.count +
              "&timestamp=" + new Date().getTime();
      },
      setProperty : function(property,value) {
         var js = this.getNewScript();
         js.src = this.cosmo.url + "?" +
              "action=post" +
              "&channel=" + this.channel +
              "&data=" + JSON.stringify([property,value]) +
              "&count=" + (this.count+1) +
              "&timestamp=" + new Date().getTime();
      },
      sync : function(update) {
            if(onCosmo) {
                 onCosmo(update);
            }
      },
      update : function(update) {
         if(update) {
            this.count++;
            this.sync(update);
         }
         this.loadScript();
      }
   };

   window.cosmo = new Cosmo(info.url);
}
var spot = cosmo.getSpot(info.room);
spot.setInfo(info.channel,info.data[1],info.data[0]);
setTimeout(
   function() {
       spot.loadScript();
   },100);

<?
         }
         break;
      case "get":
         $channel = $_REQUEST['channel'];
         $count = $_REQUEST['count'];
         $callback = stripslashes($_REQUEST['callback']);
         $channeldir = MEMDIR."/$channel";

         $next = max(2,$count+1);
         $now = time();
         $timeout = false;
         while(!file_exists("$channeldir/$next.json")) {
            usleep(50);
            if(time()-$now>10) {
                $timeout = true;
                break;
            }
         }
         $update = $timeout ? "" : file_get_contents("$channeldir/$count.json",stripslashes($data));
         if($callback) {
                header("Content-type: text/javascript");
                echo "$callback($update);";
         }
         else {
                header("Content-Type:text/plain");
                echo $update;
         }
         break;
      case "post":
         $channel = $_REQUEST['channel'];
         if($channel && ctype_alnum($channel)) {
             $data = $_REQUEST['data'];
             $count = $_REQUEST['count'];
             $channeldir = MEMDIR."/$channel";
             $filename = "$channeldir/data.json";
             if($data) {
                if(!file_exists("$channeldir/$count.json")) {
                  $fp = fopen($filename,"c+");
                  $size = filesize($filename);
                  $consolidate = flock($fp, LOCK_EX | LOCK_NB);
                  $str = $size ? fread($fp, filesize($filename)) : "";
                  if($str=="") {
                     $str = "1,{}";
                  }
                  $countstart =  substr($str,0,strpos($str,","));
                  $count = $countstart;
                }

                $next = max(2,$count+1);
                while(file_exists("$channeldir/$next.json")) {
                   $next++;
                }
                $count = $next-1;
                if($_REQUEST["timestamp"]) {
                   $array = explode("%TIMESTAMP%",$data);
                   $now = number_format(microtime(true)*1000,0,'.','');
                   $data = implode("$now",$array);
                }
                file_put_contents("$channeldir/$count.json",stripslashes($data));
                file_put_contents("$channeldir/$next.json","");
                echo $count;
             }
             
             if(!$fp) {
                 $fp = fopen($filename,"c+");
                 $consolidate = flock($fp, LOCK_EX | LOCK_NB);
             }

             if($consolidate) {
                if(!$str) {
                    $size = filesize($filename);
                    $str = $size ? fread($fp, filesize($filename)) : "";
                    if($str=="") {
                        $str = "1,{}";
                    }
                    $countstart =  substr($str,0,strpos($str,","));
                }
                ob_start();
                echo "[$str";
                for($i = $countstart; file_exists("$channeldir/$i.json"); $i++) {
                    $content = file_get_contents("$channeldir/$i.json");
                    if($content != "") {
                        echo ",$content";
                        if(!$fileage && file_exists("$channeldir/".($i-1).".json")) {
                             $fileage = time()-filemtime("$channeldir/".($i-1).".json");
                        }
                    }
                }
                echo "]";
                $topcount = $i-1;
                $str = ob_get_contents(); // get the complete string
                ob_end_clean();
                $obj = json_decode($str,true);
                $start = $obj[0];
                $len = count($obj);
                for($i = 2;$i<$len;$i++) {
                   if($obj[$i]) {
                     list($access,$value) = $obj[$i];
                     $access = explode('.',$access);
                     $leaf = &$obj[1];
                     $accesslen = count($access);
                     for($p=0;$p<$accesslen;$p++) {
                          $leafname = $access[$p];
                          if($p<$accesslen-1) {
                               if(!is_array($leaf[$leafname])) {
                                    $leaf[$leafname] = array();
                               }
                               $leaf = &$leaf[$leafname];
                          }
                     }
                     if(is_null($value)) {
                          if(isset($leaf[$leafname])) {
                               if(array_values($leaf) === $leaf)
                                    array_splice($leaf,$leafname,1);
                               else
                                    unset($leaf[$leafname]);
                          }
                     }
                     else {
                          if($leafname=="" && array_values($leaf) === $leaf)
                               $leaf[] = $value;
                          else
                               $leaf[$leafname] = $value;
                     }
                   }
                }
                $obj = $obj[1];
                $str = "$topcount,".json_encode($obj);
                fseek($fp,0);
                fwrite($fp,$str);
                fflush($fp);
                ftruncate($fp, ftell($fp));
                flock($fp, LOCK_UN);
                if($fileage>60) { //delete files older than 1 min
                    for($i = $countstart-1; $i>0 && file_exists("$channeldir/$i.json"); $i--) {
                        if(file_exists("$channeldir/$i.json"))
                           unlink("$channeldir/$i.json");
                    }
                }
            }
            fclose($fp);

         }
         break;
         default:
            echo "Cosmo - Real time communication in space";
   }
   

   function timeslot() {
       $now = microtime(true);
       return (int)($now*10) % 1000;
   }

   function destroy($folder,$destroyself) {
       if(!file_exists($folder)) {
           return;
       }
       foreach(scandir($folder) as $file) {
          if (is_dir("$folder/$file")) {
             if($file!='..' && $file!='.')
                destroy("$folder/$file",true);
          }
          else {
             unlink("$folder/$file");
          }
       }
       if($destroyself)
          rmdir ($folder);
   }
   exit();
?>